common/beerocks/bcl/unit_tests/socket_event_loop_test.cpp

```diff
@@ -0,0 +1,196 @@
+/* SPDX-License-Identifier: BSD-2-Clause-Patent
+ *
+ * SPDX-FileCopyrightText: 2020 the prplMesh contributors (see AUTHORS.md)
+ *
+ * This code is subject to the terms of the BSD+Patent license.
+ * See LICENSE file for more details.
+ */
+
+#include <bcl/beerocks_socket_event_loop.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <easylogging++.h>
+
+// Initialize easylogging++
+INITIALIZE_EASYLOGGINGPP
+
+using namespace beerocks;
+
+using ::testing::InSequence;
+using ::testing::Invoke;
+using ::testing::StrictMock;
+
+static constexpr std::chrono::milliseconds s_test_timeout(100);
+
+/**
+ * @brief Mockable event handlers class
+ */
+using EventType = std::shared_ptr<Socket>;
+using LoopType  = EventLoop<EventType>;
+
+class EventHandlersMock : public LoopType::EventHandlers {
+
+public:
+    EventHandlersMock()
+    {
+        on_read    = [&](EventType socket, LoopType &loop) { return handle_read(socket, &loop); };
+        on_write   = [&](EventType socket, LoopType &loop) { return handle_write(socket, &loop); };
+        on_timeout = [&](EventType socket, LoopType &loop) {
+            return handle_timeout(socket, &loop);
+        };
+        on_disconnect = [&](EventType socket, LoopType &loop) {
+            return handle_disconnect(socket, &loop);
+        };
+        on_error = [&](EventType socket, LoopType &loop) { return handle_error(socket, &loop); };
```

Comment:
This nicely shows why having actual methods in the EventHandlers class would be nicer :-)

You could still make a convenience subclass that has on_foo function members, if it turns out to be useful at some point. Yeah, I like that idea:

```
class EventHandler {
public:
    bool handle_read(...) = 0; // We *always* need to handle read events
    bool handle_disconnect(...) { return true; }
    ...
};

class EventHandlerFunctions {
public:
    using Handler = std::function<bool(EventType fd, EventLoopType &loop)>;
    Handler on_read = nullptr;
    ...

    bool handle_read(...) { if (on_read) return on_read(...); }
```

Of course, the EventHandlerFunctions class only needs to be added if it turns out we actually need it (which I honestly don't think we do).

Reply:
The main reason for using functions instead of a  class was to easily let the user specify in which events she was interested in. 
For example, if user is not interested in the write-event, it's enough with setting `on_write = nullptr`.

Whatever replacement of the handler functions must also consider this feature.